### PR TITLE
Query: Change JSONQuery.Path to be list of PathSegment

### LINKED ISSUE #28836: Query: Change JsonQueryExpression.Path to be a List<PathSegment>
Where PathSegment can be either 
KeyAccess which is string joined using `.`
ArrayAccess which is SqlExpression using `[]`
Or random string which we would just append, in case provide wants to add more fancier thing.

It cannot be SqlExpression even though SQL syntax allows it because neither user can write a LINQ query which accesses an arbitrary key on an object using column/parameter nor we can translate it because we need metadata of the bound property in order to materialize.

### LINKED ISSUE #28838: Remove this method and let different instances of ColumnExpression exist
https://github.com/dotnet/efcore/blob/0efc506b1103e38d8c136f9c8305368ff4bb13ef/src/EFCore.Relational/Query/JsonQueryExpression.cs#L189-L199

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #28838
Resolves #28836

### CODE REVIEW

**Comment by smitpatel** [comment]
This could be easier to review https://github.com/dotnet/efcore/pull/28888/files?w=1

2 Files which majorly changed converted to file scoped namespace.

File: src/EFCore.Relational/Query/JsonQueryExpression.cs [hunk 1]
```diff
@@ -3,275 +3,253 @@
...
+        IEntityType entityType,
+        ColumnExpression jsonColumn,
+        IReadOnlyDictionary<IProperty, ColumnExpression> keyPropertyMap,
+        Type type,
+        bool collection)
+        : this(
+              entityType,
+              jsonColumn,
+              keyPropertyMap,
+              path: new List<PathSegment> { new PathSegment("$") },
```
Comment (roji) [inline src/EFCore.Relational/Query/JsonQueryExpression.cs:39]: nit:

```suggestion
              path: new List<PathSegment> { new("$") },
```

**Review by roji — APPROVED** [review]
LGTM.

Not very important, just for my understanding: when we add JSON array indexing support (#28648), we'd augment PathSegment to support an array index in addition to a string key, right? Would the array index be a SqlExpression (e.g. some column containing the index to go into)?

**Comment by smitpatel** [comment]
> Not very important, just for my understanding: when we add JSON array indexing support (https://github.com/dotnet/efcore/issues/28648), we'd augment PathSegment to support an array index in addition to a string key, right? Would the array index be a SqlExpression (e.g. some column containing the index to go into)?

Yes. Exactly that is how we will add support.

**Pull request merged**